****🌟Exercise 1: Introduction to Data Analysis****

**1. What Is Data Analysis?**

Data analysis is the systematic process of collecting, cleaning, organizing, transforming, and examining data with a sole purpose of discovering useful information, patterns, and meaningful insights that can support better decisions. For example, a business may have thousands of records containing customer purchases. By analyzing this data, the business can determine which products are most popular, when customers are most likely to buy, and which products are underperforming. Instead of relying on assumptions, managers can use evidence from the data to make informed decisions.
Data analysis can also be divided into different approaches. Descriptive analysis examines what happened, diagnostic analysis investigates why something happened, predictive analysis estimates what may happen in the future, and prescriptive analysis helps determine what action should be taken.

**2. Why Is Data Analysis Important in Modern Contexts?**

Ultimately, the value of data does not come simply from having large amounts of information. Its value comes from being able to analyze that information effectively and transform it into reliable insights that can improve decisions and solve real-world problems.

One of the main benefits of data analysis is better decision-making. Instead of making decisions based primarily on assumptions or intuition, organizations can use evidence and measurable trends. For example, a company can analyze sales data to determine which products should receive more investment.
Data analysis also helps organizations identify problems and improve efficiency. An organization can examine its operations to identify delays, unnecessary costs, declining performance, or areas where resources are being wasted. In healthcare, for example, analyzing patient and population data can support treatment decisions, disease monitoring, and long-term planning.
Another important benefit is prediction and planning. Organizations can use historical data together with statistical models and machine learning to forecast demand, identify risks, and anticipate future trends.

**3. Three Areas Where Data Analysis Is Applied Today**

**In Environmental management and Agriculture:** Environmental managers and farmers can analyze information about rainfall, temperature, soil conditions, water quality, biodiversity, land use, and vegetation to understand environmental changes and support sustainable resource management. For example, satellite imagery and GIS data can be analyzed to monitor changes in forest cover, identify areas affected by drought, assess agricultural productivity, and monitor environmental degradation. Combining these data sources with statistical analysis and machine learning can help decision-makers develop more effective environmental and agricultural strategies.

**In Hospitals and healthcare** organizations collect information from electronic health records, medical tests, patient surveys, wearable devices, and public health databases. Analyzing this information can help healthcare professionals identify health risks, support diagnosis and treatment decisions, monitor diseases, and allocate resources. At a population level, health authorities can analyze disease patterns and vaccination data to understand public health trends and plan interventions. For example, predictive models can analyze patient information to identify people who may be at high risk of developing complications. This allows healthcare professionals to be proactive than reactive.

**In Businesses** data analysis is used to understand customers, monitor performance, forecast sales, manage inventory, and improve marketing strategies. For example, an online retailer can analyze customers' purchasing histories to identify popular products and recommend relevant products to individual customers. Entrepreneurs can also use data analysis to identify market opportunities. By studying customer demographics, purchasing behavior, competitors, and market trends, entrepreneurs can make better decisions about what products or services to develop.


**🌟Exercise 2: Dataset Loading and Initial Analysis**

Dataset #1: How Much Sleep Do Americans Really Get?

In [22]:
# Import the datasets from the required data sources

import pandas as pd

# Load the data directly into a DataFrame
sleeping_time_df = pd.read_csv("American_sleeping_time.csv")

# View the first few rows
sleeping_time_df.head(10)


,index,Year,Period,Avg hrs per day sleeping,Standard Error,Type of Days,Age Group,Activity,Sex
0,0,2003,Annual,8.57,0.018,All days,15 years and over,Sleeping,Both
1,1,2004,Annual,8.55,0.026,All days,15 years and over,Sleeping,Both
2,2,2005,Annual,8.62,0.023,All days,15 years and over,Sleeping,Both
3,3,2006,Annual,8.63,0.024,All days,15 years and over,Sleeping,Both
4,4,2007,Annual,8.57,0.024,All days,15 years and over,Sleeping,Both
5,5,2008,Annual,8.60,0.023,All days,15 years and over,Sleeping,Both
6,6,2009,Annual,8.67,0.023,All days,15 years and over,Sleeping,Both
7,7,2010,Annual,8.67,0.024,All days,15 years and over,Sleeping,Both
8,8,2011,Annual,8.71,0.026,All days,15 years and over,Sleeping,Both
9,9,2012,Annual,8.73,0.026,All days,15 years and over,Sleeping,Both


The dataset contains information about Americans’ sleeping time across different years from 2003 to 2017. It includes variables such as year, period, average hours of sleep, standard error, activity type, age group, and sex. The data shown focuses on people aged 15 years and over, for both males and females, and records average sleeping hours annually.
For example, in 2003, Americans aged 15+ slept an average of 8.57 hours per day, while in 2007 the average was 8.57 hours.

Dataset #2: Global Trends in Mental Health Disorder

In [23]:
# Load the data directly into a DataFrame
mental_health_df = pd.read_csv("Mental_health_disorder.csv")

# View the first few rows
mental_health_df.head(10)


C:\Users\LOMBE\AppData\Local\Temp\ipykernel_30884\651800925.py:2: DtypeWarning: Columns (5,6) have mixed types. Specify dtype option on import or set low_memory=False.
  mental_health_df = pd.read_csv("Mental_health_disorder.csv")


,index,Entity,Code,Year,Schizophrenia (%),Bipolar disorder (%),Eating disorders (%),Anxiety disorders (%),Drug use disorders (%),Depression (%),Alcohol use disorders (%)
0,0,Afghanistan,AFG,1990,0.16056,0.697779,0.101855,4.828830,1.677082,4.071831,0.672404
1,1,Afghanistan,AFG,1991,0.160312,0.697961,0.099313,4.829740,1.684746,4.079531,0.671768
2,2,Afghanistan,AFG,1992,0.160135,0.698107,0.096692,4.831108,1.694334,4.088358,0.670644
3,3,Afghanistan,AFG,1993,0.160037,0.698257,0.094336,4.830864,1.705320,4.096190,0.669738
4,4,Afghanistan,AFG,1994,0.160022,0.698469,0.092439,4.829423,1.716069,4.099582,0.669260
5,5,Afghanistan,AFG,1995,0.160076,0.698695,0.09098,4.828337,1.728112,4.104207,0.668746
6,6,Afghanistan,AFG,1996,0.160249,0.698914,0.089709,4.828083,1.737643,4.107500,0.667727
7,7,Afghanistan,AFG,1997,0.160554,0.699127,0.088372,4.827726,1.746891,4.110834,0.666220
8,8,Afghanistan,AFG,1998,0.160931,0.699372,0.08733,4.826971,1.756963,4.114438,0.664676
9,9,Afghanistan,AFG,1999,0.161311,0.699674,0.086267,4.826413,1.770791,4.117633,0.663428


The dataset contains information about Global trnd in mental health across different years from 1990 to 2017. The general variables are: Entity(countries), entity code and year. The health variables in percentage are : schizophrenia, depression, and disorders (bipolar, eating, anxiety, alcohol and drug use)


Dataset #3: Credit Card Approvals

In [24]:
# Load the data directly into a DataFrame
credit_card_df = pd.read_csv("crx.csv")

# View the first few rows
credit_card_df.head()


,b,30.83,0,u,g,w,v,1.25,t,t.1,01,f,g.1,00202,0.1,+
0,a,58.67,4.460,u,g,q,h,3.04,t,t,6,f,g,00043,560,+
1,a,24.50,0.500,u,g,q,h,1.50,t,f,0,f,g,00280,824,+
2,b,27.83,1.540,u,g,w,v,3.75,t,t,5,t,g,00100,3,+
3,b,20.17,5.625,u,g,w,v,1.71,t,f,0,f,s,00120,0,+
4,b,32.08,4.000,u,g,m,v,2.50,t,f,0,t,g,00360,0,+


In [25]:
# Load the data directly into a DataFrame
clean_dataset_df = pd.read_csv("clean_dataset.csv")

# View the first few rows
clean_dataset_df.head()


,Gender,Age,Debt,Married,BankCustomer,Industry,Ethnicity,YearsEmployed,PriorDefault,Employed,CreditScore,DriversLicense,Citizen,ZipCode,Income,Approved
0,1,30.83,0.000,1,1,Industrials,White,1.25,1,1,1,0,ByBirth,202,0,1
1,0,58.67,4.460,1,1,Materials,Black,3.04,1,1,6,0,ByBirth,43,560,1
2,0,24.50,0.500,1,1,Materials,Black,1.50,1,0,0,0,ByBirth,280,824,1
3,1,27.83,1.540,1,1,Industrials,White,3.75,1,1,5,1,ByBirth,100,3,1
4,1,20.17,5.625,1,1,Industrials,White,1.71,1,0,0,0,ByOtherMeans,120,0,1


**🌟Exercise 3: Identifying Data Types**

For the datasets from the previous exercise, categorize each column of it as either quantitative or qualitative and explain your reasoning.

| Dataset | Column | Type | Reasoning |
|---|---|---|---|
| American_sleeping_time.csv | index | Qualitative | Row identifier, not a measured quantity |
| American_sleeping_time.csv | Year | Qualitative (ordinal) | Time label, not a quantity to average |
| American_sleeping_time.csv | Period | Qualitative | Text label (only "Annual") |
| American_sleeping_time.csv | Avg hrs per day sleeping | Quantitative | Continuous measured value (hours) |
| American_sleeping_time.csv | Standard Error | Quantitative | Continuous statistical measure |
| American_sleeping_time.csv | Type of Days | Qualitative | Categorical text |
| American_sleeping_time.csv | Age Group | Qualitative (ordinal) | Binned, ordered categories |
| American_sleeping_time.csv | Activity | Qualitative | Text label (only "Sleeping") |
| American_sleeping_time.csv | Sex | Qualitative | Categorical (Both/Men/Women) |
| Mental_health_disorder.csv | index | Qualitative | Row identifier |
| Mental_health_disorder.csv | Entity | Qualitative | Country/region name |
| Mental_health_disorder.csv | Code | Qualitative | ISO country code |
| Mental_health_disorder.csv | Year | Qualitative (ordinal) | Time label |
| Mental_health_disorder.csv | Schizophrenia (%) | Quantitative | Continuous prevalence rate |
| Mental_health_disorder.csv | Bipolar disorder (%) | Quantitative | Continuous prevalence rate |
| Mental_health_disorder.csv | Eating disorders (%) | Quantitative | Continuous prevalence rate |
| Mental_health_disorder.csv | Anxiety disorders (%) | Quantitative | Continuous prevalence rate |
| Mental_health_disorder.csv | Drug use disorders (%) | Quantitative | Continuous prevalence rate |
| Mental_health_disorder.csv | Depression (%) | Quantitative | Continuous prevalence rate |
| Mental_health_disorder.csv | Alcohol use disorders (%) | Quantitative | Continuous prevalence rate |
| clean_dataset.csv | Gender | Qualitative | Binary code representing a category |
| clean_dataset.csv | Age | Quantitative | Continuous measured value |
| clean_dataset.csv | Debt | Quantitative | Continuous monetary amount |
| clean_dataset.csv | Married | Qualitative | Binary flag |
| clean_dataset.csv | BankCustomer | Qualitative | Binary flag |
| clean_dataset.csv | Industry | Qualitative | Text category |
| clean_dataset.csv | Ethnicity | Qualitative | Text category |
| clean_dataset.csv | YearsEmployed | Quantitative | Continuous duration |
| clean_dataset.csv | PriorDefault | Qualitative | Binary flag |
| clean_dataset.csv | Employed | Qualitative | Binary flag |
| clean_dataset.csv | CreditScore | Quantitative | Numeric score with meaningful magnitude |
| clean_dataset.csv | DriversLicense | Qualitative | Binary flag |
| clean_dataset.csv | Citizen | Qualitative | Text category |
| clean_dataset.csv | ZipCode | Qualitative | Numeric-looking but a location code |
| clean_dataset.csv | Income | Quantitative | Continuous monetary amount |
| clean_dataset.csv | Approved | Qualitative | Binary outcome label (target) |


🌟Exercise 4: Exploring Data Types\
Learn about different types of data in data analysis.

In [ ]:
# Install the Kaggle API client
!pip install kaggle


In [20]:
import pandas as pd
import requests
import zipfile
import io

url = "https://github.com/devtlv/Datasets-DA-Bootcamp-2-/raw/refs/heads/main/Week%204%20-%20Data%20Understanding/W4D2%20-%20Data%20U.%20and%20pre-processing/Iris%20Dataset.zip"

response = requests.get(url)

with zipfile.ZipFile(io.BytesIO(response.content)) as z:
    print(z.namelist())

['Iris Dataset/Iris.csv']


In [22]:
with zipfile.ZipFile(io.BytesIO(response.content)) as z:
    df = pd.read_csv(z.open("Iris Dataset/Iris.csv"))

df.head()

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,1,5.1,3.5,1.4,0.2,Iris-setosa
1,2,4.9,3.0,1.4,0.2,Iris-setosa
2,3,4.7,3.2,1.3,0.2,Iris-setosa
3,4,4.6,3.1,1.5,0.2,Iris-setosa
4,5,5.0,3.6,1.4,0.2,Iris-setosa



| Column | Type | Why |
|---|---|---|
| Id | Qualitative (Nominal) | Unique identifier/label for each row; numbers are arbitrary, not measurements. |
| SepalLengthCm | Quantitative (Continuous) | Numeric measurement in cm; supports averages and other arithmetic. |
| SepalWidthCm | Quantitative (Continuous) | Continuous numeric measurement of width in cm. |
| PetalLengthCm | Quantitative (Continuous) | Continuous numeric measurement of petal length. |
| PetalWidthCm | Quantitative (Continuous) | Continuous numeric measurement of petal width. |
| Species | Qualitative (Nominal) | Category label (setosa/versicolor/virginica) with no natural numeric order. |


🌟 Exercise 6: Identifying Data Types\
Below are various data sources. Identify whether each one is an example of structured or unstructured data.


| Data Source | Structured / Unstructured | Why |
|---|---|---|
| Company's financial reports in an Excel file | Structured | Data sits in rows and columns with defined fields (dates, amounts, categories), easy to search, sort, and analyze directly. |
| Photographs uploaded to a social media platform | Unstructured | Images have no predefined fields or tabular format; extracting meaning requires specialized processing like computer vision. |
| A collection of news articles on a website | Unstructured | Free-form text with no fixed schema separating things like sentiment, topic, or key facts. |
| Inventory data in a relational database | Structured | Organized into predefined tables with rows and columns (SKU, quantity, price, location), following a strict schema. |
| Recorded interviews from a market research study | Unstructured | Audio/spoken content with no fixed structure; requires transcription and further processing to extract usable information. |


🌟 Exercise 7: Transformation Exercise


| Source | Best fit | Why (in one line) |
|---|---|---|
| Travel blog posts | **Crowdsourcing** | Authentic, personal experience is the core value |
| Customer service audio | **Simulation** | Enables controlled, PII-safe, rare-scenario coverage |
| Handwritten brainstorm notes | **Crowdsourcing** | Diversity of handwriting + genuine creative content |
| Cooking tutorial video | **Simulation** | Precise labeling and safe coverage of rare/dangerous actions |

🌟 Exercise 8 : Import a file from Kaggle

In [23]:
import pandas as pd
import requests
import zipfile
import io

url = "https://github.com/devtlv/Datasets-DA-Bootcamp-2-/raw/refs/heads/main/Week%204%20-%20Data%20Understanding/W4D3%20-%20Importing%20Data,%20Exporting%20D/train.zip"

response = requests.get(url)

with zipfile.ZipFile(io.BytesIO(response.content)) as z:
    print(z.namelist())

['train.csv']


In [24]:
with zipfile.ZipFile(io.BytesIO(response.content)) as z:
    df = pd.read_csv(z.open("train.csv"))

df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


🌟 Exercise 9 : Export a dataframe to excel format and JSON format.

In [36]:
import numpy as np
import pandas as pd
from faker import Faker

# Create a Faker object
fake = Faker()

# List of fruits
fruits = [
    "Apple",
    "Banana",
    "Orange",
    "Mango",
    "Pineapple",
    "Watermelon",
    "Grapes",
    "Strawberry",
    "Papaya",
    "Avocado"
]

# Generate 100 synthetic sales records
data = []

for _ in range(100):
    product_name = fake.random.choice(fruits)
    sale_amount = np.random.uniform(5, 10)
    date_of_sale = fake.date_between(start_date='-2y', end_date='today')
    customer_age = np.random.randint(18, 70)

    data.append([
        product_name,
        sale_amount,
        date_of_sale,
        customer_age
    ])

# Convert the data into a Pandas DataFrame
sales_df = pd.DataFrame(
    data,
    columns=['Product Name', 'Sale Amount', 'Date of Sale', 'Customer Age']
)

# Display the first 5 rows
print(sales_df.head())

  Product Name  Sale Amount Date of Sale  Customer Age
0   Strawberry     5.520945   2025-05-12            59
1       Grapes     7.022179   2026-03-09            37
2       Orange     6.049416   2026-06-04            34
3   Watermelon     6.529257   2026-07-27            56
4        Apple     6.915961   2025-11-29            68


In [34]:
#Export to JSON
sales_df.to_json("sales_dataframe.json")

#Export to xlsx
sales_df.to_excel("sales_dataframe.xlsx", sheet_name = "Sales_Records")

🌟 Exercise 10: Reading JSON Data

In [25]:
import pandas as pd
import requests
import zipfile
import io

url = "https://github.com/devtlv/Datasets-DA-Bootcamp-2-/raw/refs/heads/main/Week%204%20-%20Data%20Understanding/W4D3%20-%20Importing%20Data,%20Exporting%20D/posts.zip"

response = requests.get(url)

with zipfile.ZipFile(io.BytesIO(response.content)) as z:
    print(z.namelist())

['posts.json']


In [26]:
with zipfile.ZipFile(io.BytesIO(response.content)) as z:
    df = pd.read_json(z.open("posts.json"))

df.head()

,userId,id,title,body
0,1,1,sunt aut facere repellat provident occaecati e...,quia et suscipit\nsuscipit recusandae consequu...
1,1,2,qui est esse,est rerum tempore vitae\nsequi sint nihil repr...
2,1,3,ea molestias quasi exercitationem repellat qui...,et iusto sed quo iure\nvoluptatem occaecati om...
3,1,4,eum et est occaecati,ullam et saepe reiciendis voluptatem adipisci\...
4,1,5,nesciunt quas odio,repudiandae veniam quaerat sunt sed\nalias aut...
